### DISCUSSION TITLE
`createdIds` example

### DISCUSSION THREAD

**Discussion by piboistudios** [jmapio/jmap#396] [Q&A] [answered]
Hi,

I see `createdIds` referenced a few times but in the docs it isn't very clearly specified and there is no example. Also, in every existing JMAP implementation I've found, it is basically just passed through on the client side (haven't seen any server side implementations that implement it either)

Namely, there is no example of how a batch requests may reference a createdId.

**Can someone provide an example using `createdIds`?**

Also, while I understand a table/map lookup for IDs yields better performance than reference lookups, I don't get what this feature does that using a `ResultReference` cannot solve. 

E.g. **how would referencing a createdId be different than just using a `ResultReference`?**

**Comment by zvasilev** [comment]
Here is an example of an email sent request/response. 
The creation ID is an ephemeral random string to match the server response. 
See in the request.json file, the "create" property has a creation ID object "aefe," and in the response, there is a property "created" with the object "aefe."

[request.json](https://github.com/jmapio/jmap/files/14553577/request.json)
[response.json](https://github.com/jmapio/jmap/files/14553575/response.json)

**Reply by piboistudios** [reply]
Thank you for the example!

Although, I am actually referring to the optional `createdIds` property of [requests](https://jmap.io/spec-core.html#the-request-object)

Is that property entirely unnecessary? It seems that way from your example. E.g. it seems like even without the createdIds property, the server will still map creation id references to the actual ID of the underlying resource.

**Comment by neilj** [comment]
There seem to be 2 different questions here. The first one is why do creation id references exist in addition to ResultReferences. The answer to that is because they can be used in different places. You can only use a ResultReference to do a simple syntactical replacement of a top-level argument to a function, while a creation id can be used anywhere the real id can be used.

The 2nd question is why can you pass in a `createdIds` property with the Request. This makes life much, much easier when you are proxying JMAP calls through to potentially multiple servers. This may require splitting up what was a single request into multiple requests to different backends, and you can just pass the createdIds property along, which means creation references still work without the proxy having to resolve them all in between making each request.

**Reply by piboistudios** [reply]
Well, what I am wondering mostly is what createdIds would look like on a request.

It's a map (`Id[Id]`)  but the client would only have the creation IDs they generate, so does that mean the values in the map can be any value, as the server will overwrite it with the actual corresponding `Id` in response?

e.g.
in the request:
```
"createdIds": {
  "A1": "foo"
}
```

and in the response:
```
"createdIds": {
  "A1": "someId123"
}
```

**Reply by piboistudios** [reply]
> There seem to be 2 different questions here.

Yes, I bolded them both.

> The 2nd question is why can you pass in a createdIds property with the Request. 

That's not the 2nd question I asked (or a question I asked at all, and also not even remotely close). I bolded the questions I asked.  

Please don't mark questions as closed immediately after responding without even answering them if you actually want people to make new JMAP implementations. It's quite hostile and only seems to indicate you _don't_ want to help.

Based on the way you rearranged my questions (and completely changed one), it seems like you took the second question personally and then made the other question personal too. This isn't a critique. I was genuinely curious... But I'll leave that curiosity to others, this put a bad taste in my mouth, especially with you being the primary maintainer.

**Comment by zvasilev** [comment, marked as answer]
Assume createdIds as a temporary ID generated by the client app.
If you pay a bit more attention to the response.json, the client ID "aefe" is not replaced by the server ID "S1416". They both exist in the response.
But if you make more than one create object, this is the way to know **how to map both ids**!

Each JMAP call after that should use the server ID for "update" or "delete," etc.

This is the way for the server to return a stable immutable ID and not rely on some ill generated client ID.
Then, all subsequent calls could be stable as both client and server use one stable ID. Just take a look on IMAP Plus, etc extensions mess.

Yes, createdIds could be anything by RFC.
By providing createdIds, it is possible to bind another JMAP call without the need to wait for the server to return its immutable message ID.
